### PR TITLE
initial commit to enable amx

### PR DESCRIPTION
**Pull request by ziyizhang-1**
As mentioned in [x86_amx_intrinsics](https://github.com/rust-lang/rust/issues/126622)

Completed AMX Intrinsics:
`amx-tile`
- _tile_loadconfig
- _tile_storeconfig
- _tile_loadd
- _tile_release
- _tile_stored
- _tile_stream_loadd
- _tile_zero
`amx-int8`
- _tile_dpbssd
- _tile_dpbsud
- _tile_dpbusd
- _tile_dpbuud
`amx-bf16`
- _tile_dpbf16ps
`amx-fp16`
- _tile_dpfp16ps
`amx-complex`
- _tile_cmmimfp16ps
- _tile_cmmrlfp16ps

### CODE REVIEW

File: crates/core_arch/src/x86_64/mod.rs
Comment (sayantn) [inline crates/core_arch/src/x86_64/mod.rs:38]: Can't this be `Derive`d?

File: crates/stdarch-test/src/lib.rs
Comment (sayantn) [inline crates/stdarch-test/src/lib.rs:207]: These syscalls are required only on the linux kernel, and you can use the `syscall` crate as a `dev-dependency`. Windows by default enables amx support. Also, this should be in `core_arch/src/x86/test.rs`

File: crates/core_arch/src/x86_64/amxtile.rs
Comment (sayantn) [inline crates/core_arch/src/x86_64/amxtile.rs:34]: You can use `llvm.x86.sttilecfg` here

File: crates/core_arch/src/x86_64/amxtile.rs
Comment (sayantn) [inline crates/core_arch/src/x86_64/amxtile.rs:18]: `llvm.x86.ldtilecfg` here

**Comment by sayantn** [comment]
Can you please run `cargo test -p stdarch-verify` with the `PRINT_MISSING_LISTS_MARKDOWN` constant set to `true`? This would update the `missing-x86.md` file to reflect the new added intrinsics. Also, please do not edit the `x86-intel.xml` file manually. Download the latest from Intel's website (I think v3.6.9 currently) and replace this

**Review by sayantn — CHANGES_REQUESTED** [review]

File: crates/core_arch/src/x86_64/mod.rs
Comment (ziyizhang-1) [reply crates/core_arch/src/x86_64/mod.rs]: Thx for the advice, implement several basic traits for __tilecfg via `derive`

File: crates/stdarch-test/src/lib.rs
Comment (ziyizhang-1) [reply crates/stdarch-test/src/lib.rs]: Since only a few testcases of amx intrinsics require this function and probably not expose to users, so maybe it is too expensive to introduce the syscall crate

File: crates/core_arch/src/x86_64/amxtile.rs
Comment (ziyizhang-1) [reply crates/core_arch/src/x86_64/amxtile.rs]: replaced~

**Comment by sayantn** [comment]
Even though windows doesn't need the syscall, you still need to have the function to compile. Maybe add an empty `init_amx` function with `#[cfg(all(target_arch = "x86_64", windows))]`

**Comment by ziyizhang-1** [comment]
> Even though windows doesn't need the syscall, you still need to have the function to compile. Maybe add an empty `init_amx` function with `#[cfg(all(target_arch = "x86_64", windows))]`

yep, that's a good workaround to prevent repeated condition check

**Comment by ziyizhang-1** [comment]
> Can you please run `cargo test -p stdarch-verify` with the `PRINT_MISSING_LISTS_MARKDOWN` constant set to `true`? This would update the `missing-x86.md` file to reflect the new added intrinsics. Also, please do not edit the `x86-intel.xml` file manually. Download the latest from Intel's website (I think v3.6.9 currently) and replace this.
> 
> Also, there are very few amx intrinsics, so they can be implemented in a single `amx.rs` file instead of 5 files

updated `x86-intel.xml` to v3.6.9, please help to review the change

File: crates/core_arch/src/x86_64/mod.rs
Comment (sayantn) [inline crates/core_arch/src/x86_64/mod.rs:25]: Can you add a method in `__tilecfg` that takes the `pub` fields as input and returns the correctly configured cfg (as the `reserved` fields are private, this struct can't be constructed outside `core_arch`)?

File: crates/stdarch-verify/tests/x86-intel.rs
Comment (sayantn) [inline crates/stdarch-verify/tests/x86-intel.rs:718]: This is not needed, you are using `i64` as `size_t`

File: crates/core_arch/src/x86/test.rs
Comment (sayantn) [inline crates/core_arch/src/x86/test.rs:185]: See [Intel's sample code](https://github.com/intel/AMX-TMUL-Code-Samples/blob/main/src%2Ftest-amxtile.c#L85-L99). They use only one syscall to directly set xtiledata without checking for xtilecfg (as xtiledata requires xtilecfg, that call would simply fail if xtilecfg is not enabled). Also, as you are only using it in one file (`x86_64/amx.rs`), it makes sense to move these functions to the `tests` module of amx.rs

**Review by sayantn — CHANGES_REQUESTED** [review]

**Comment by sayantn** [comment]
The `x86_64-unknown-linux-gnu-emulated` CI runs all the tests (using Intel SDE), so your implementation amx-complex and amx-fp16 has been tested

File: crates/core_arch/src/x86_64/mod.rs
Comment (Amanieu) [reply crates/core_arch/src/x86_64/mod.rs:25]: The GCC and Clang intrinsics don't define a `__tilecfg` type and just use `void *`. We shouldn't be inventing our own type here.

File: crates/core_arch/src/x86/test.rs
Comment (Amanieu) [inline crates/core_arch/src/x86/test.rs:147]: This should specifically be `cfg(target_os = "linux")`. The syscalls are specific to Linux and won't work on other OSes like BSD.

File: crates/stdarch-test/src/lib.rs
Comment (Amanieu) [reply crates/stdarch-test/src/lib.rs]: It's fine to add dev-dependencies. They are only used when testing this crate in CI.

File: crates/core_arch/src/x86_64/mod.rs
Comment (ziyizhang-1) [reply crates/core_arch/src/x86_64/mod.rs:25]: Oh, thx for the notification, I did not notice the function signature of the ldtilecfg and sttilecfg in llvm which can be found [here](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/IR/IntrinsicsX86.td). So I will make the `ldtilecfg` to accept `*const i8` instead of `*const __tilecfg` and similar to `sttilecfg`

File: crates/stdarch-verify/tests/x86-intel.rs
Comment (ziyizhang-1) [reply crates/stdarch-verify/tests/x86-intel.rs:718]: I will remove that from 32-bit as I do not think amx can be supported on those processors

Comment (sayantn) [reply crates/stdarch-verify/tests/x86-intel.rs:718]: yes, amx is x64-only

File: crates/core_arch/src/x86/test.rs
Comment (ziyizhang-1) [reply crates/core_arch/src/x86/test.rs:185]: The first syscall is arch_prctl(ARCH_GET_XCOM_PERM) which is a read-only operation and has 2~3x less overhead compared with the latter syscall arch_prctl(ARCH_REQ_XCOMP_PERM), which might execute a write operation. The `xtiledata` flag only needs to be set once for a process during its lifetime. So we can apply the first "read" syscall to check if there is a need to follow up another "request" syscall in case the xtiledata has already been turned on.

File: crates/core_arch/src/x86_64/mod.rs
Comment (ziyizhang-1) [reply crates/core_arch/src/x86_64/mod.rs:25]: I think the __tilecfg is necessary to call amx intrinsics, so if we do not define it here, users might be forced to do so. How about we provide a usable definition here? Since the interface accepts `*const i8`, it is also okay for users to pass their self-defined struct to the function.

File: crates/stdarch-verify/tests/x86-intel.rs
Comment (sayantn) [inline crates/stdarch-verify/tests/x86-intel.rs:719]: `size_t` should be an unsigned type, preferably `usize` (you can define the LLVM intrinsic signature with `stride` as `usize`, LLVM doesn't care about the signedness or type of integer arguments, just their bit width)

File: crates/core_arch/Cargo.toml
Comment (sayantn) [inline crates/core_arch/Cargo.toml:28]:
```suggestion
[target.'cfg(all(target_arch = "x86_64", target_os = "linux"))'.dev-dependencies]
syscalls = { version = "0.6.18", default-features = false }
```

File: crates/stdarch-verify/tests/x86-intel.rs
Comment (ziyizhang-1) [reply crates/stdarch-verify/tests/x86-intel.rs]: Added `usize` as a primitive in x86-intel, please check if that what you expect

File: crates/core_arch/src/x86_64/amx.rs
Comment (sayantn) [inline crates/core_arch/src/x86_64/amx.rs:58]: Normally in Rust, const generics are named in all-caps

File: crates/core_arch/src/x86_64/amx.rs
Comment (sayantn) [inline crates/core_arch/src/x86_64/amx.rs:240]: Why is this lint turned off? Seems like stagnant code

File: crates/core_arch/src/x86_64/amx.rs
Comment (ziyizhang-1) [reply crates/core_arch/src/x86_64/amx.rs]: renamed

File: crates/core_arch/src/x86_64/amx.rs
Comment (ziyizhang-1) [reply crates/core_arch/src/x86_64/amx.rs]: Good question, those are redundant, removed

**Review by sayantn — APPROVED** [review]

**Comment by sayantn** [comment]
@Amanieu should we keep the `__tilecfg` type? It indeed makes for a nice abstraction (all C programs that use amx have to define a `struct` or use an array because C lacks this type)

**Comment by Amanieu** [comment]
I don't think we should be inventing our own API in this crate, we should strictly follow the vendor's API. What if Intel decides to add its own `__tilecfg` type in the future that is incompatible with ours? I think it is best to leave this to an external crate. That crate can also deal with other AMX-related issues such as the syscalls needed for initialization.

File: crates/core_arch/src/x86_64/amx.rs
Comment (Amanieu) [inline crates/core_arch/src/x86_64/amx.rs:11]: For void pointers in Rust we generally use `*mut u8`. This is what the allocator API uses for example. At the LLVM level there is no distinction between `i8` and `u8` which is why LLVM intrinsics are always show taking `i8`.

Comment (ziyizhang-1) [reply crates/core_arch/src/x86_64/amx.rs:11]: make sense, let's make it more idiomatic

**Comment by ziyizhang-1** [comment]
I understood the concern raised by @Amanieu; there is indeed some probability that Intel will change the definition of the tile configuration in the future, for example, increasing/decreasing the number of tiles, which can lead to a change of the config layout. Anyway, I just moved `__tilecfg` to the tests mod, and considering the current feature `x86_amx_intrinsics` is still unstable, any different points of view can be welcomed and discussed.

**Comment by Amanieu** [comment]
Everything else looks good, so this can be merged once the `i32` issue is fixed.

**Pull request merged**
